ts/services/ReportService.ts

```diff
@@ -1,36 +1,150 @@
+            const elementFilter = elementsInScope.content.map( (x) => x.element_id );
+            const groupedCost = await this.getGroupedCost(config, logger, report, elementFilter);
+            if (config.format === 'text') {
+              logger.info(groupedCost.content.sort((rpt1, rpt2) => {
+                return rpt1.display_category.localeCompare(rpt2.display_category);
+              }).map((rpt) => {
+                return rpt.display_category + ' (Cost: ' + rpt.total_cost + ')';
+              }));
+            }
+
+            // json
+            if (config.format === 'json') {
+              logger.info(JSON.stringify(groupedCost, null, 2));
+            }
+          }
+        }
+      }
+    } catch (e) {
+      logger.error('There was an error getting ec2 cost data: ' + e);
+    }
+  }
+
+  private async getReports(config, logger): Promise<Report[]> {
+    try {
+      const response = await request({
+        auth: {
+          pass: config.password,
+          user: config.username
+        },
+        json: true,
+        uri: config.endpoint + '/reports'
+      });
+      return response.reports;
+    } catch (e) {
+      logger.error('There was an error getting the reports: ' + e);
+      return [];
+    }
+  }
+
+  private async getElementsInScope(config, logger, report: Report): Promise<ReportContent> {
+
+    const reportScope = new ReportScope('elementsInScope', {
+      elementTypes: ['EC2', 'WINSVR', 'SERVER'],
+      endDate: this.getEndDate(report.endDate, config.timerange),
+      startDate: this.getStartDate(report.endDate, config.timerange)
+    },
+    config.rowlimit);
+    return this.doReportContentPost(config, logger, reportScope, report);
+
+  }
+
+  private async getGroupedCost(config, logger, report: Report, elementIds: string[]): Promise<ReportContent> {
+
+    const reportScope = new ReportScope('groupedCost', {
+      activeQuantityAgg: 'sum',
+      categoriseBy: 'category',
+      elementFilter: elementIds,
+      endDate: this.getEndDate(report.endDate, config.timerange),
+      groupBy: config.groupby,
+      groupKey: 'attribute=' + config.groupbykey,
+      instanceTypeKey: 'instanceType',
+      service: 'EC2',
+      startDate: this.getStartDate(report.endDate, config.timerange)
+    },
+    config.rowlimit);
+    return this.doReportContentPost(config, logger, reportScope, report);
+
+  }
+
+  private async doReportContentPost(config, logger, bodyObject, report) {
+    try {
+      const response = await request.post({
+        auth: {
+          pass: config.password,
+          user: config.username
+        },
+        body: bodyObject,
+        json: true,
+        uri: config.endpoint + '/reports/content/' + report.id
+      });
+      return response;
+    } catch (e) {
+      logger.error('There was an error fetching the report content: ' + e);
+    }
+  }
+
+  private getStartDate(endDate: Date, timerange: string) {
+
+    return timerange === 'lastMonth' ?
```

Comment:
This is a very long (double) ternary operation. Can you break it up into something more readable? E.g.

```typescript
if (timerange === 'lastMonth') {
    return moment(endDate).utc().subtract(1, 'months').startOf('month').toISOString();
}
if (timerange === 'monthToDate') {
    return moment(endDate).utc().startOf('month').toISOString();
}
return moment(endDate).toISOString();
```

Or use a switch statement, which this seems like it is.

Reply:
I prefer early returns over long, broken up lines. I don't have strong enough grounds for a switch if you like it, though, so I'm fine with you leaving it.